include/exadg/structure/spatial_discretization/operator.h

```diff
@@ -379,12 +393,25 @@ class Operator : public dealii::Subscriptor, public Interface::Operator<Number>
   std::shared_ptr<dealii::FiniteElement<dim>> fe;
   dealii::DoFHandler<dim>                     dof_handler;
   dealii::AffineConstraints<Number>           affine_constraints;
-  // constraints for mass operator (we use a separate AffineConstraints object, because we do not
-  // apply constraints from Dirichlet boundary conditions here)
-  dealii::AffineConstraints<Number> constraints_mass;
 
-  std::string const dof_index                = "dof";
-  std::string const dof_index_mass           = "dof_mass";
+  // To treat inhomogeneous Dirichlet BCs correctly in the context of matrix-free operator
+  // evaluation using dealii::MatrixFree/FEEvaluation, we need a separate AffineConstraints
+  // object containing only periodicity and hanging node constraints.
+  // When using the standard AffineConstraints object including Dirichlet boundary conditions,
+  // inhomogeneous boundary data would be ignored by dealii::FEEvaluation::read_dof_values().
+  // While dealii::FEEvaluation::read_dof_values_plain() would take into account inhomogeneous
+  // Dirichlet data using the standard AffineConstraints object, hanging-node constraints would
+  // not be resolved correctly.
```

Comment:
Let me lay out two alternative designs to this comment as a discussion before we finalize the design for now, not necessary to change anything here:
- `read_dof_values_plain` could be kept by calling `AffineConstraints::distribute` on the object containing only hanging nodes and periodicity constraints, but I think the way we choose now is clearer in terms of the evaluators.
- Of course it would be even nicer if we could encode two different sets of constraints into a single `dof_index` of `MatrixFree` and call `read_dof_values()` and `read_dof_values_unconstrained` (not reflecting the more strict category), but I do not think we are there yet. I am not even sure if it can be realized without ambiguity from the perspective of ExaDG in all cases we have.

As I said, I do not believe we want to go for a different route, but these are the two other reasonable options I see, so I wanted to state them once again for the completeness of discussion.